FreeTubeApp/FreeTube
# Implement Stylelint linting of CSS logical properties

# Implement Stylelint linting of CSS logical properties




## Pull Request Type


## Related issue
See [comment](https://github.com/FreeTubeApp/FreeTube/pull/3970#issuecomment-1697564131)https://github.com/FreeTubeApp/FreeTube/pull/3970#issuecomment-1697564131

## Description
This change implements the Stylelint plugin [stylelint-use-logical-spec](https://github.com/Jordan-Hall/stylelint-use-logical-spec) with all rules enabled except for requiring `float: inline-start` and `float: inline-end`, which as of today is still an experimental feature in Chrome.

To prevent current issues blocking contributors going forward, this PR also enacts this rule throughout FreeTube. This mostly just changes all CSS properties with `height` and `width` to become `block-size` and `inline-size` respectively, which is only functionally relevant for different [`writing-mode`s](https://developer.mozilla.org/en-US/docs/Web/CSS/writing-mode) (note that the default `horizontal-tb` is the only `writing-mode` currently supported by FreeTube, but that is subject to change in the future).

This PR enables Stylelint for `npm run lint` and `npm run lint-fix`, including minor modification of the other scripts to make space for this. This PR also runs Stylelint on pre-commit. This was never enabled beforehand due to concerns about which other rules should be needed. As a result, this PR removes the other plugins and rules, and they can be added back as necessary in their own tested PRs.

## Screenshots 
![Screenshot_20230910_161625](https://github.com/FreeTubeApp/FreeTube/assets/84899178/6d4f3a9a-ebc4-41ef-912e-59dcd936419e)

## Testing 
1. Add a non-logical property (e.g., `padding-right: 55px`) to a CSS file (e.g., `App.vue`).
2. Save the file and try to commit the change.
3. Encounter pre-commit hook notifying you of the issue with the non-logical property.
4. Run `npm run lint` to see the same assessment as in 3.
5. Run `npm run lint-fix` to see the issue automatically resolved by the linter.
6. Commit the fixed change; you should encounter no issues.
7. (Optional) run other slightly modified commands to ensure they work correctly (`eslint-lint`, `eslint-lint-fix`, `lint-style-fix`)

## Desktop

- **OS:** OpenSUSE Tumbleweed
- **OS Version:** 2023xxxx
- **FreeTube version:** 0.19.0

## Additional context
Given that `transform` and `float` are still vulnerable to introducing directionality concerns, code contributors and reviewers must take special care to [use these CSS horizontal directionality variables going forward](https://github.com/FreeTubeApp/FreeTube/blob/c96f8c58c915feb23355a227e15f311ed5f96531/src/renderer/themes.css#L1008-L1010). We currently only care about the x-dimension for right-to-left language support. Code contributors and reviewers must also take care to use [logical properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_logical_properties_and_values) inside `style` blocks and properties in `.js` and `.vue` files. Currently, the [stylelint-use-logical-spec](https://github.com/Jordan-Hall/stylelint-use-logical-spec) plugin is not capable of detecting these instances. We should probably add this note to the FreeTube Docs as well.
